### PR TITLE
add the option to use a `FORCE_CUDA` to force cuda installation on docker

### PR DESCRIPTION
**Pull request by bendidi**
cf [#167](https://github.com/facebookresearch/maskrcnn-benchmark/issues/167) 

this seems to be the best workaround

### CODE REVIEW

**Comment by miguelvr** [comment]
it is better to add `FORCE_CUDA` as a docker build arg so that it is still possible to build with CPU support only

**Comment by miguelvr** [comment]
You can keep it as a default set to true. 

Plus, the documentation in `INSTALL.md` needs to be updated accordingly

File: docker/Dockerfile [hunk 1]
```diff
@@ -32,7 +33,6 @@ RUN conda install -y ipython
 RUN pip install ninja yacs cython matplotlib opencv-python
 
 # Install PyTorch 1.0 Nightly
-ARG CUDA
```
Comment (miguelvr) [inline docker/Dockerfile]: This needs to be here

File: docker/Dockerfile [hunk 2]
```diff
@@ -47,8 +47,8 @@ RUN git clone https://github.com/cocodataset/cocoapi.git \
  && python setup.py build_ext install
 
 # install PyTorch Detection
-RUN git clone https://github.com/facebookresearch/maskrcnn-benchmark.git \
+RUN git clone https://github.com/bendidi/maskrcnn-benchmark.git \
```
Comment (miguelvr) [inline docker/Dockerfile]: ???

Comment (bendidi) [reply docker/Dockerfile]: just for testing, on my remote gpu

File: docker/Dockerfile (hunk 1 above)
Comment (bendidi) [reply docker/Dockerfile]: fixed

File: docker/Dockerfile (hunk 2 above)
Comment (bendidi) [reply docker/Dockerfile]: fixed

File: docker/Dockerfile [hunk 3]
```diff
@@ -1,5 +1,6 @@
 ARG CUDA="9.0"
 ARG CUDNN="7"
+ARG FORCE_CUDA="1"
```
Comment (miguelvr) [inline docker/Dockerfile]: this will go out of context as soon as the `FROM` is called, so it needs to be further down the line.

File: docker/Dockerfile [hunk 4]
```diff
@@ -48,8 +48,8 @@ RUN git clone https://github.com/cocodataset/cocoapi.git \
  && python setup.py build_ext install
 
 # install PyTorch Detection
-RUN git clone https://github.com/bendidi/maskrcnn-benchmark.git \
+RUN git clone https://github.com/facebookresearch/maskrcnn-benchmark.git \
  && cd maskrcnn-benchmark \
- && FORCE_CUDA=1 python setup.py build develop
```
Comment (miguelvr) [inline docker/Dockerfile]: please do it like this:

```
# install PyTorch Detection
ARG FORCE_CUDA="1"
ENV FORCE_CUDA=${FORCE_CUDA}
RUN git clone https://github.com/facebookresearch/maskrcnn-benchmark.git \
 && cd maskrcnn-benchmark \
 && python setup.py build develop
```

build args are limited to the scope of the build and are not environment variables, they are intended to change the behavior of the build.

**Comment by miguelvr** [comment]
@bendidi can you also modify the instructions in `INSTALL.md`? Thanks!

File: demo/README.md [hunk 5]
```diff
@@ -38,7 +38,8 @@ docker run --rm -it \
     -v /tmp/.X11-unix:/tmp/.X11-unix \
     --device=/dev/video0:/dev/video0 \
     --ipc=host maskrcnn-benchmark \
-    python demo/webcam.py --min-image-size 300
+    python demo/webcam.py --min-image-size 300 \
```
Comment (miguelvr) [inline demo/README.md]: is this change necessary?

Comment (bendidi) [reply demo/README.md]: yes the default is `../configs/caffe2/e2e_mask_rcnn_R_50_FPN_1x_caffe2.yaml` , while when running from docker and from `maskrcnn-benchmark` folder  it should be `configs/caffe2/e2e_mask_rcnn_R_50_FPN_1x_caffe2.yaml`

**Comment by bendidi** [comment]
> @bendidi can you also modify the instructions in `INSTALL.md`? Thanks!

There are no modifications to be added to `install.md`, the installation instructions stays the same. this is just a fix for the installation on docker problem and doesn't impact the process

File: demo/README.md (hunk 5 above)
Comment (miguelvr) [reply demo/README.md]: alright, thanks!

**Comment by miguelvr** [comment]
> > @bendidi can you also modify the instructions in `INSTALL.md`? Thanks!
> 
> There are no modifications to be added to `install.md`, the installation instructions stays the same. this is just a fix for the installation on docker problem and doesn't impact the process

there are changes because there is a new build arg in the docker

File: demo/README.md (hunk 5 above)
Comment (fmassa) [reply demo/README.md]: Note that the demo won't work if you run it from `maskrcnn-benchmark` folder due to the local includes, so this breaks in that case

Comment (fmassa) [reply demo/README.md]: for example, trying to `import predictor` won't work if done from `maskrcnn-benchmark`, and it is explicitly mentioned in the readme that you should cd to `demo` first, so I'd vote for removing this change

Comment (bendidi) [reply demo/README.md]: running the current command: 
```
docker run --rm -it \
    -e DISPLAY=${DISPLAY} \
    --privileged \
    -v /tmp/.X11-unix:/tmp/.X11-unix \
    --device=/dev/video0:/dev/video0 \
    --ipc=host maskrcnn-benchmark \
    python demo/webcam.py --min-image-size 300
```
throws the error `FileNotFoundError: [Errno 2] No such file or directory: '../configs/caffe2/e2e_mask_rcnn_R_50_FPN_1x_caffe2.yaml'`

this change is just to specify path to config file in this particular case: 
```
docker run --rm -it \
    -e DISPLAY=${DISPLAY} \
    --privileged \
    -v /tmp/.X11-unix:/tmp/.X11-unix \
    --device=/dev/video0:/dev/video0 \
    --ipc=host maskrcnn-benchmark \
    python demo/webcam.py --min-image-size 300 \
    --config-file configs/caffe2/e2e_mask_rcnn_R_50_FPN_1x_caffe2.yaml
```

it should not break any imports in the demo folder

**Comment by bendidi** [comment]
@miguelvr @fmassa good to merge ?

**Comment by miguelvr** [comment]
Okay by me, but @fmassa has to merge

File: demo/README.md (hunk 5 above)
Comment (fmassa) [reply demo/README.md]: I just noticed that this is for the `docker` example.

I have never tried using it with docker, but I know for sure that without docker running the webcam demo code from `maskrcnn-benchmark` folder will not work, you need to be in the `demo` folder.

But maybe some paths are setup differently in the dockerfile.

Comment (bendidi) [reply demo/README.md]: wierd, works fine for me outside of `demo` folder (not in docker) using python 3.6 :

```
python demo/webcam.py --min-image-size 300 --config-file configs/caffe2/e2e_mask_rcnn_R_50_FPN_1x_caffe2.yaml MODEL.DEVICE cpu
```
